rfc/20241118-module-vars-and-outputs-deprecation.md

```diff
@@ -0,0 +1,178 @@
+deprecation of module input variables and outputs.
+
+Currently, module authors either keep backward compatiblity or put deprecation notices in the release notes. Both options are
+not ideal for the end-user. Therefore, it would be useful if OpenTofu could notify module users about the deprecation of input 
+variables or outputs as defined by module authors.
+
+> [!NOTE]
+> Module deprecation as a whole is a similar problem, however it is addressed on the OpenTofu Registry level.
+
+## Proposed Solution
+
+OpenTofu should allow module authors to mark input variables and outputs as deprecated with a warning message alongside.
+Module users have to see a respective warning message in cases, when deprecated input variable or output is used in their
+configuration.
+
+> [!NOTE]
+> The result of `tofu validate`, `tofu plan`, `tofu apply` and `tofu test` should include deprecation warnings.
+
+### User Documentation
+
+Module authors should be able to use a new field named `deprecated` in both `variable` and `output` blocks. This field 
+should accept a required string value for a warning message:
+
+```hcl
+variable "this_is_my_variable" {
+  type        = string
+  description = "This is a variable for the old way of configuring things."
+  deprecated  = "This variable will be removed on 2024-12-31. Use that_is_my_variable instead."
+}
+
+output "this_is_my_output" {
+  value       = some_resource.some_name.some_field
+  description = "This is an output for the old way of using things."
+  deprecated  = "This output will be removed on 2024-12-31. Use that_is_my_output instead."
+}
+```
+
+#### Deprecated module variable warning
+
+If module caller specify non-null value for a deprecated variable, OpenTofu should produce a deprecation warning as 
+specified by module author:
+
+```hcl
+│ Warning: The variable "this_is_my_variable" is marked as deprecated by module author.
+│ 
+│   on mod/main.tf line 9, in module call "mod":
+│    9:     this_is_my_variable = "something"
+│ 
+│ This variable will be removed on 2024-12-31. Use that_is_my_variable instead.
+```
+
+> [!NOTE]
+> It is possible for module user to specify a null value for a deprecated variable and receive no warnings. OpenTofu treats
+> variables with explicit null values the same way as if it was ommited completely. This is not an ideal, however an idiomatic
+> OpenTofu language approach. See `null` values note [in the docs](https://opentofu.org/docs/language/expressions/types/#types).
+
+#### Deprecated module output warning
+
+If module caller has references to deprecated module outputs in their configuration, OpenTofu should produce a
+deprecation warning as specified by module author:
+
+```hcl
+│ Warning: The output "this_is_my_output" is marked as deprecated by module author.
+│ 
+│   on main.tf line 9, in resource "some_resource" "some_name":
+│    9:     some_field = module.mod.this_is_my_output
+│ 
+│ This output will be removed on 2024-12-31. Use that_is_my_output instead.
+```
```

Comment:
What do you think ought to happen if the calling module were written like this instead?

```hcl
module "mod" {
  # ...
}

locals {
  mod = module.mod
}

resource "some_resource" "some_name" {
  # ...

  some_field = local.mod.this_is_my_output
}
```

This is the most straightforward kind of indirection, but really all sorts of things are possible here because `module.mod` is just a normal value of an object type, and so the author of the calling module is allowed to perform arbitrary transformations on it before actually accessing the value that was originally in the `this_is_my_output` attribute of that object. :confounded: 

---

Another tricky variation is any case where the `module` block is using `count`/`for_each`. In that case the reference is likely to appear as something like `module.mod[each.key].this_is_my_output`, which appears to OpenTofu's graph builder only as a reference to `module.mod` because the rest of the expression is dynamic and so HCL's `Expression.Variables` function can't "see" the `.this_is_my_output` part as a static reference.